miniupnpc/minissdpc.c

```diff
@@ -463,14 +463,25 @@ static int upnp_gettimeofday(struct timeval * tv)
 	tv->tv_usec = (ts % 1000) * 1000;
 	return 0; /* success */
 #elif defined(CLOCK_MONOTONIC_FAST) || defined(CLOCK_MONOTONIC)
-	struct timespec ts;
-	int ret_code = clock_gettime(UPNP_CLOCKID, &ts);
-	if (ret_code == 0)
+#if defined(__APPLE__)
```

Comment:
when `clock_gettime()` is not available, `CLOCK_MONOTONIC_FAST` and `CLOCK_MONOTONIC` are not defined, right ? so it is already falling back to `gettimeofday()`, right ?

Reply:
@miniupnp:
AFAIK, When compiling with a modern macOS SDK:
- `CLOCK_MONOTONIC` and `clock_gettime` are both defined, regardless of `MACOSX_DEPLOYMENT_TARGET`
- They are guarded with availability notations, signifying that if you set (for example) a `MACOSX_DEPLOYMENT_TARGET` < 10.12, they will not *work* on prior platforms (and will be weak-linked).
- Instead of doing things like manually checking if `clock_gettime` is `NULL` at runtime (which will not silence the warnings), the recommended solution is using `__builtin_available`, which handles only executing the branch if the code is running on supported versions of Apple platforms.

So, no, if compiling on a modern macOS SDK: `clock_gettime` and `CLOCK_MONOTONIC` are always defined in the headers, regardless of whether they will actually be available at run-time. Hence the warning, and - while I haven't tested this specific case - likely a crash when running on earlier versions of Apple platforms (without a change like the above).

Reply:
is  `__builtin_available() ` when compiling on every Apple platform / compiler ? I'm afraid it will prevent buildin with gcc.

I would prefer #undef CLOCK_MONOTONIC_FAST and CLOCK_MONOTONIC if MACOSX_DEPLOYMENT_TARGET < 10.12

Reply:
@miniupnp:

`__builtin_available` is indeed Clang-specific, but works for every Apple platform (assuming Clang / Xcode). It also does some "fancy" things depending on what the minimum deployment target is, to avoid any penalty unless necessary (IIRC).

> I would prefer #undef CLOCK_MONOTONIC_FAST and CLOCK_MONOTONIC if MACOSX_DEPLOYMENT_TARGET < 10.12

AFAIK, `MACOSX_DEPLOYMENT_TARGET` isn't passed as a define to source code. It's used, as an environment variable, to default various flags sent to the compiler / linker. (Those flags can also be specified directly - ex `-mmacosx-version-min=10.12`.)

However if you'd like to make this smart for clang, and "working" for gcc, it could be modified to something like:
```
#if defined(__APPLE__)
#if defined(__clang__)
	if (__builtin_available(macOS 10.12, iOS 10.0, tvOS 10.0, watchOS 3.0, *)) {
#else
	if (clock_gettime != NULL) {
#endif // defined(__clang__)
```

My understanding is this will do "the proper thing" for clang / Xcode (with all the advantages that come with it, including smart behavior depending on deployment target and suppressing the warning / error), and will avoid calling a weak-linked `clock_gettime` on other compilers (which _should_ be set to NULL, and presumably will lack the warning / error).

Reply:
I have updated the PR with my suggested strategy above to maintain non-clang compatibility.